### PR TITLE
Fix negative delta arguments

### LINKED ISSUE #26887: Getting negative `delta` values for `_process`
**Godot version:**
Godot 3.1. RC 1

**OS/device including version:**
Windows 10 Home 64-bit, Geforce GTX 860m

**Issue description:**
I already found some similar issue, but it was related to android only. Therefore I named the issue similar but the error also seems to occur on windows machines.

I noticed my script crashes because of a division by zero (Stripped down to relevant part):

```gdscript
func _process(delta):
	if abs(angle) > PI * 4 * delta:
		thisRotation = current.slerp(target, max_radians_delta / abs(angle))
	else:
		thisRotation = current.slerp(target, 1)
```

For my case, angle = 0. If delta is negative the first code path is executed and results in a division by zero error. This happens sometimes when I start my game and do not move my character.

If this is relevant: I use the bullet physics with KinematicBody and StaticBodyies.

**Minimal reproduction project:**
I would provide the project if not easy to reproduce, I think a simple print if delta < 0 is sufficient.

### PR DESCRIPTION
**Pull request by zmanuel**
Three attack points, all after the regular calculations:
1. Prevent negative physics timestep counts. They could occur if
   physics_jtter_fix is changed at runtime.
2. idle_step is not allowed to go below one tick. That could
   happen on physics_jitter_fix changes or heavily fluctuating
   performance.
3. Prevent that the idle_step modification breaks the promise
   that Engine.get_physics_interpolation_fraction() is between
   0 and 1 by doing more physics steps than the base system wants.

Fixes #26887

### CODE REVIEW

File: main/main_timer_sync.cpp [hunk 1]
```diff
@@ -30,6 +30,11 @@
 
 #include "main_timer_sync.h"
 
+#ifdef DEBUG_ENABLED
+// enable to get diagnostics from here
+// #define SYNC_TIMER_DEBUG_ENABLED
+#endif
```
Comment (aaronfranke) [inline main/main_timer_sync.cpp]: This does nothing.

File: main/main_timer_sync.cpp [hunk 2]
```diff
@@ -172,9 +191,37 @@ MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iteratio
 	// last clamping: make sure time_accum is between 0 and p_frame_slice for consistency between physics and idle
 	ret.clamp_idle(idle_minus_accum, idle_minus_accum + p_frame_slice);
 
+	// all the operations above may have turned ret.idle_step negative or zero, keep
+	// at least one tick
+	const float min_step = 1 / 1000000.0;
```
Comment (aaronfranke) [inline main/main_timer_sync.cpp]: Might be better to write as `1e-6`.

**Review by aaronfranke — DISMISSED** [review]

File: main/main_timer_sync.cpp (hunk 1 above)
Comment (lawnjelly) [reply main/main_timer_sync.cpp]: > This does nothing.

Did you mean having the compile time diagnostics? While in many cases I would agree, as someone else who works on the timing code, I personally would find this more useful than to exclude it.

File: main/main_timer_sync.cpp (hunk 2 above)
Comment (zmanuel) [reply main/main_timer_sync.cpp]: @aaronfranke 1E-6 would be my preference as well, but the codebase has a different opinion.
```
$ grep "1E-6" * -R | wc -l
4
$ grep "1000000\.0" * -R | wc -l
30
```
The three uses of 1E-6 are all collision penetration tolerances, the fourth was from a binary, where all conversions from ticks to seconds I found are done in the ticks/1000000.0 form.

The SYNC_TIMER_DEBUG_ENABLED guards many runtime checks later in the code. The commented out line is there so if further problems appear, I can quickly enable the extra diagnostics again. They're not enabled in general even for debug builds for various reasons:
The warnings are just diagnostics that the new code has kicked in and prevented errors. The test case triggers them all, so in general, they're useless noise.
Some of the CRASH_CONDs have floating point comparisons in them, and you know how those are. They might trigger false positive crashes if they're just slightly violated (which still would be fine).
None of the CRASH_CONDs are of the 'we can't continue if this condition holds' type, they just indicate something has gone wrong in this part of the code that should be looked at.

I can rework those to only emit warnings in debug mode for everyone if the expected constraints are not met. That way, we get some diagnostics that something is going wrong that can be easily traced to the source without risking spurious crashes.

Comment (zmanuel) [reply main/main_timer_sync.cpp]: Thinking again, taking one tick as the minimum output timestep here is wrong. A single tick is small enough to cause trouble in careless _process functions, for example if they use
`1 - cos(delta*omega)`
(which they should not!) with some omega in the range of one. A better minimal output timestep is probably a quarter of the input timestep.

Comment (lawnjelly) [reply main/main_timer_sync.cpp]: In practice I don't think you can guess in advance how users are going to use this delta (and thus there is no 'safe' value). I added something like this on a similar PR on a suggestion, but in retrospect really I think the best answer is to require all code to work when passed a zero delta.

A suggestion - we may be able to enforce this by passing a test zero delta at some point / points during runtime. This would ensure that any such bugs in core and client code were fixed sharpish.

Comment (zmanuel) [reply main/main_timer_sync.cpp]: > A suggestion - we may be able to enforce this by passing a test zero delta at some point / points during runtime. This would ensure that any such bugs in core and client code were fixed sharpish.

That sounds like a good idea to me. Bad client code should fail fast. If code has problems with zero delta, it will likely also misbehave for very small delta. And worst case, the code runs into an error and aborts (only that one _process call). What is the expected result from that? No change, which is what is supposed to happen with zero delta.
I'd be for adding a _process(0) call once to any freshly added object, randomly as first to 100ths call or similar. Debug mode only, maybe. And then change the two bits of code that guarantee a minimal delta (yours in the basic timer query and mine here) to just guarantee non-negative delta.

Comment (avencherus) [reply main/main_timer_sync.cpp]: > A suggestion - we may be able to enforce this by passing a test zero delta at some point / points during runtime.

That's a very good idea.  These kinds of rare mistakes can be nightmarish to track down.

**Comment by Faless** [comment]
Probably also fixes: #25166

File: main/main_timer_sync.cpp [hunk 3]
```diff
@@ -143,9 +166,15 @@ MainFrameTime MainTimerSync::advance_core(float p_frame_slice, int p_iterations_
...
 // calls advance_core, keeps track of deficit it adds to animaption_step, make sure the deficit sum stays close to zero
 MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iterations_per_second, float p_idle_step) {
+	if (p_idle_step <= 0) {
+		WARN_PRINT_ONCE("p_idle_step not positive");
+	}
+
 	if (fixed_fps != -1)
 		p_idle_step = 1.0 / fixed_fps;
 
+	const float min_output_step = p_idle_step > 0 ? p_idle_step * .25 : 1E-6;
```
Comment (nathanwfranke) [inline main/main_timer_sync.cpp]: Shouldn't the minimum just be zero? Also, maybe this line should be moved down to where the if statement is

Comment (zmanuel) [reply main/main_timer_sync.cpp]: If we think that _process() should be able to handle zero as delta argument, yes, min_output_step should just be zero. The unexpectedly crashing code from #26887 would handle it fine.
 In the current form, it probably shouldn't be moved further down because the next line changes p_idle_step, possibly making it negative; that would then lead to the minimal minimal step of 1E-6, which I'm not overly fond of.

I count that as two votes in favor of allowing delta=0 as argument to _process. Anyone against it?

Comment (nathanwfranke) [reply main/main_timer_sync.cpp]:
```
func _process(delta):
    translation.x += 100.0 * delta # Delta being 0 is valid
```

**Comment by zmanuel** [comment]
I removed the positive limit on output deltas and only guarantee they are non-negative now. The PR is updated with the squashed result. The test project (https://github.com/godotengine/godot/issues/26887#issuecomment-578441699) then complains about lots of zero deltas, of course, but no negative ones.

The previous code is still here: 7d29221d363194bd140f732d31149ced2f7addcf

**Comment by nathanwfranke** [comment]
Why would there be zero delta? Is the frame rate higher than 100,000?

**Comment by zmanuel** [comment]
No, but the incoming delta times fluctuate by much more than the average delta and also by more than 1/physics_fps. The system here tries to keep the outgoing deltas stable within the given bounds by physics_jitter_fix, and that sometimes means the "game clock" (sum of outgoing deltas) is ahead of the real clock (sum of incoming deltas), this will happen if the framerate increased from one moment to the next. And then again sometimes, the system decides that being ahead was a mistake and that it should resync game and real clock while the current incoming delta is very small, smaller than the difference between the two times. The bug was that it would do so by letting the game clock snap back, creating a negative delta. Now it just stops for a bit.
The previous revision had it slow down instead, which was a bit iffy because it involved two magic numbers, the absolute minimum delta (for your >1E+6 FPS case, for example) and the slowdown factor.

File: main/main_timer_sync.cpp (hunk 1 above)
Comment (aaronfranke) [reply main/main_timer_sync.cpp:36]: What I mean is that, as-is, this code does nothing. Perhaps line 35 should be uncommented.

Comment (zmanuel) [reply main/main_timer_sync.cpp:36]: Perhaps. But as explained in the slightly [wrong place further down](https://github.com/godotengine/godot/pull/35617#discussion_r371824901), `SYNC_TIMER_DEBUG_ENABLED` just guards debug output statements that I'd rate as noise for anyone not explicitly dealing with problems in this area. Some of them do trigger often at program start or when breakpoints are set/hit in the editor.

Two alternative suggestions:

1. Change the comment to `// uncomment the following line to enable extended diagnostics from this source file` to make the purpose clearer
2. Remove the block and all the logging guarded by `SYNC_TIMER_DEBUG_ENABLED`. It's not that useful and yeah, code commented out and left around "in case we need it later" is not pretty

File: main/main_timer_sync.cpp (hunk 3 above)
Comment (nathanfranke) [reply main/main_timer_sync.cpp]: > Everyone expects delta to be change-in-time-since-last-process. It is physically impossible for a perfectly accurate timer to say no time has passed, as calling functions always takes time

It is physically impossible to store a delta time less than the float epsilon, and I would vote in favor of more accurate long-term timers. If you add `delta` to a time variable every frame, and the script is executing faster than `1/epsilon` frames per second, then the timer would run too fast. Regardless, one would need hundreds of thousands of FPS to get a zero delta in the first place (possibly millions).

Comment (zmanuel) [reply main/main_timer_sync.cpp]: For the timers Godot uses, granularity is bigger than the CPU clock (I think it was 10^6 ticks per second?). So yes, it is physically possible to get a measurement of zero for the passed time since the last frame, however improbable that is. If you're dividing by the delta argument to update, you're making two mistakes: Doing a potentially failing operation without checking, and doing a numerically unstable operation for small delta arguments. Yes, that these error sources exist is a surprise for novices. Ideally, we'd throw a delta=0 and delta=10^-6 argument at them in debug mode every once in a while to make everyone get used to the possibility; it's hard do so that without disrupting game flow, however.

**Comment by zmanuel** [comment]
So, discussion here is a biiiit slow and it would be nice to get this fix in, because the current situation of negative delta is just no good. I've amended the fix here:
https://github.com/zmanuel/godot/tree/negative-delta-fix-unsquashed2
Two changes:
- the "in case we need it" debug code is removed; that has caused some confusion and is of little value
- the minimal timestep is now the measured timestep divided by 8. Not 0, that's apparently not universally loved. It's still a hardcoded number, but pretty much all values between 2 and 1024 have the same effect; it's not a magic value that needs fine tuning. So I'm personally OK with it.

It's also rebased to current master. As the name says, the changes are not squashed to a single commit yet.

What's the procedure there? Obviously, squash. But then, do I just force push it to update this MR? Or do I make a new MR, refering back here? Ordinarily, I'd update, but between the original submission and now was the call to resubmit all MRs anyway. That call also said old MRs would be deactivated/rejected/closed, which hasn't happened here.

**Comment by akien-mga** [comment]
You can force push to update this PR, the announcement we made that old PRs might be closed has not really been enforced, we're doing our best to try to integrate whatever was submitted in the last and keep being updated by their author.

File: main/main_timer_sync.cpp (hunk 3 above)
Comment (Jimmio92) [reply main/main_timer_sync.cpp]: I don't see what all the fuss is. No gain will come from receiving a time based event with no time having passed. I for one cannot believe that this simple thing is so debated and left this way.

I have written toy engines myself in C and GL and have had to fix jitter and the like, and none of it was ever because no time passed between frames. This sounds 100% like covering up the real problem: Bad time source.

On Win32, QueryPerformanceCounter provides for nanosecond accuracy on most hardware that I tested. This guarantees that the second query will be after the first, and there's no way that, given the amount of work Godot must do, for a delta of 0 or less to EVER appear.

Even if it did appear, you're doing work based on time. Whatever that work is, you'd do none of it if delta was zero. Wasted cycles, wasted energy. Just throw out the process event if <= 0, seriously. It's not like anything is going to break... no time passed, remember? And the next tiny slice of time will get the process call, so again... no problems, but it's fixed. No potential gachas for newbies who (imo correctly) assume that time always marches forward. No chance of causing any issues because of a dropped _process, because it'll happen again *extremely* soon.

Comment (zmanuel) [reply main/main_timer_sync.cpp]: You are 100% correct: All of this IS a way to compensate for a bad timer source. The problem isn't the timer resolution, though, it's that the time is taken on the CPU. In the best case (single threaded operation), when all the commands have been passed to the GPU. In the worst case (multithread), when all the scene data has been passed to the rendering thread. If you want stable frame time measurements, you should take the time when the GPU is starting to deliver the newly rendered frame to the screen. Alas, in GLES, that seems to be impossible. In Desktop OpenGL and Vulkan, you can use fences and queries to determine when the GPU flipped the buffers N frames ago. It's a bit difficult because it's asynchronous, but doable. Patches probably welcome.

The entire purpose of the file that is getting patched here is to take the imperfect timer measurements and derive stable process calls. Focus is on getting consistent number of physics steps per render frame. Without the code in this file, Godot would sometimes, even in perfect 60FPS VSync locked conditions, do jittery updates where each frame, anything between 0 and 2 physics updates were performed, because the imperfect timer measurements would lie just a little bit on either side of a threshold. Averaging over N frames makes these occurrences rarer, but does not fix them.

The negative timesteps that are corrected in this patch aren't coming from the timer itself, they're a bad output of the correction. Essentially, it's overshooting a bit one frame, then tries to backpedal. Maybe it would be possible to just prevent the overshooting; I don't think so, because that's what the system has to do to function correctly when frame times decrease. I opted for the post-correction also because that guarantees there are no side effects, it's easier to prove that the result will be correct, and it'll stay correct even if the core gets changed.

Just not calling _process() if the argument would be 0 is also not such a good option. There are three possible expectations here:
1. The sum of delta arguments to _process() is as close to the total program runtime as possible. Networked games rely on that.
2. the argument to _process() should be positive. That is what you want, even though it was never really guaranteed (just extremely unlikely to be false).
3. _process is called for every rendered frame. That is [documented](https://docs.godotengine.org/en/stable/getting_started/step_by_step/scripting_continued.html?highlight=_process#processing).
Best possible conformance with expectation 1, in the situation covered here, requires to either violate 2 or 3. You get screamed at by someone no matter what :)

The current, updated patch opts to violate expectation 1 a little to save 2 and 3. But not more than it is allowed to; the limit on real time/game time deviation still is less than physics_jitter_fix/physics_fps. I personally would prefer to break bad client code, but do so early; I don't really see a way to do that well, though, and don't want that to be the hill this bugfix dies on.

File: main/main_timer_sync.cpp [hunk 4]
```diff
@@ -147,10 +160,16 @@ MainFrameTime MainTimerSync::advance_core(float p_frame_slice, int p_iterations_
...
+	if (p_idle_step <= 0) {
+		WARN_PRINT_ONCE("p_idle_step not positive");
+	}
+
 	if (fixed_fps != -1) {
 		p_idle_step = 1.0 / fixed_fps;
 	}
 
+	const float min_output_step = p_idle_step > 0 ? p_idle_step / 8 : 1E-6;
+
```
Comment (lawnjelly) [inline main/main_timer_sync.cpp:172]:
```
min_output_step = p_idle_step / 8;
min_output_step = MAX(min_output_step, 1E-6);
```
Might be better, otherwise you could get small values of p_idle_step reaching zero.

Comment (zmanuel) [reply main/main_timer_sync.cpp:173]: Right, that's safer. Unless we actually get a million FPS :)

File: main/main_timer_sync.cpp [hunk 5]
```diff
@@ -105,6 +105,15 @@ MainFrameTime MainTimerSync::advance_core(float p_frame_slice, int p_iterations_
 		}
 	}
 
+#ifdef DEBUG_ENABLED
+	if (max_typical_steps < 0) {
+		WARN_PRINT_ONCE("max_typical_steps is negative");
```
Comment (Calinou) [inline main/main_timer_sync.cpp:110]:
```suggestion
		WARN_PRINT_ONCE("`max_typical_steps` is negative. This could hint at an engine bug or system timer misconfiguration.");
```

File: main/main_timer_sync.cpp [hunk 6]
```diff
@@ -105,6 +105,15 @@ MainFrameTime MainTimerSync::advance_core(float p_frame_slice, int p_iterations_
...
+	}
+	if (min_typical_steps < 0) {
+		WARN_PRINT_ONCE("min_typical_steps is negative");
```
Comment (Calinou) [inline main/main_timer_sync.cpp:113]:
```suggestion
		WARN_PRINT_ONCE("`min_typical_steps` is negative. This could hint at an engine bug or system timer misconfiguration.");
```

File: main/main_timer_sync.cpp [hunk 7]
```diff
@@ -147,10 +160,17 @@ MainFrameTime MainTimerSync::advance_core(float p_frame_slice, int p_iterations_
 
 // calls advance_core, keeps track of deficit it adds to animaption_step, make sure the deficit sum stays close to zero
 MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iterations_per_second, float p_idle_step) {
+	if (p_idle_step <= 0) {
+		WARN_PRINT_ONCE("p_idle_step not positive");
```
Comment (Calinou) [inline main/main_timer_sync.cpp:164]:
```suggestion
		WARN_PRINT_ONCE("`p_idle_step` is not strictly positive. This could hint at an engine bug or system timer misconfiguration.");
```

File: main/main_timer_sync.cpp [hunk 8]
```diff
@@ -177,9 +197,37 @@ MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iteratio
...
+
+	if (time_accum > p_frame_slice) {
+		const int extra_physics_steps = floor(time_accum * p_iterations_per_second);
+		time_accum -= extra_physics_steps * p_frame_slice;
+		ret.physics_steps += extra_physics_steps;
+	}
+
+#ifdef DEBUG_ENABLED
+	if (time_accum < -1E-7) {
+		WARN_PRINT_ONCE("time_accum negative");
```
Comment (Calinou) [inline main/main_timer_sync.cpp:224]:
```suggestion
		WARN_PRINT_ONCE("`time_accum` is negative. This could hint at an engine bug or system timer misconfiguration.");
```

File: main/main_timer_sync.cpp [hunk 9]
```diff
@@ -177,9 +197,37 @@ MainFrameTime MainTimerSync::advance_checked(float p_frame_slice, int p_iteratio
...
+	}
+
 	// restore time_accum
 	time_accum = ret.idle_step - idle_minus_accum;
 
+	// forcing ret.idle_step to be positive may trigger a violation of the
+	// promise that time_accum is between 0 and p_frame_slice
+#ifdef DEBUG_ENABLED
+	if (time_accum < -1E-7) {
+		WARN_PRINT_ONCE("time_accum negative");
```
Comment (Calinou) [inline main/main_timer_sync.cpp:212]:
```suggestion
		WARN_PRINT_ONCE("`time_accum` is negative. This could hint at an engine bug or system timer misconfiguration.");
```

**Review by Calinou — CHANGES_REQUESTED** [review]
Should be good to merge after applying suggestions and squashing commits together.

File: main/main_timer_sync.cpp (hunk 8 above)
Comment (zmanuel) [reply main/main_timer_sync.cpp]: Done, squashed and rebased to current master (adapting to name changes there).

File: main/main_timer_sync.cpp (hunk 3 above)
Comment (jordo) [reply main/main_timer_sync.cpp]: > The negative timesteps that are corrected in this patch aren't coming from the timer itself, they're a bad output of the correction. Essentially, it's overshooting a bit one frame, then tries to backpedal. Maybe it would be possible to just prevent the overshooting; I don't think so, because that's what the system has to do to function correctly when frame times decrease. I opted for the post-correction also because that guarantees there are no side effects, it's easier to prove that the result will be correct, and it'll stay correct even if the core gets changed.

I'll just comment here, to show my support.  But the above comment is really is bang on.  This really needs to get clamped in all circumstances and no circumstance pass a negative delta to an iteration step.  If the timing PURELY used the jittery CPU timer sampler, it sill would guarantee >= delta times.  If any 'smoothing' or 'physics fix' post timer fix is causing a negative delta, this needs to get clamped to > 0 as soon as possible.  

With regards to violation of expectation 1, it should be ok if there is small amount of accumulated delta drift across frame/iteration boundaries, but not across the long-term time accumulator itself.  Which is think is covered here.  It's correc that we can't have the delta arguments to _process() continuously drift away from total program runtime.  Variance between iterations is fine, but long term definitely not fine.  As far as I can tell from this PR, all time is accounted for and accumulated in the end.

**Comment by jordo** [comment]
> So, discussion here is a biiiit slow and it would be nice to get this fix in, because the current situation of negative delta is just no good.

I'm hoping to get some of the reviewers eyes on this soon.

**Comment by lawnjelly** [comment]
We discussed this PR a bit today in rocket chat:
https://chat.godotengine.org/channel/devel/thread/ECRSuitLBW9BxSiTJ

and the general consensus was at least me and reduz were happy to try this out in the 3.4 betas. That is assuming it has been tested and appears to work as well as can be tested prior to a wider rollout. If there are any problems I'm guessing they will appear pretty soon.

In the long run we might do some improvements to the timing code / refactoring but this should be good to fix the bugs in the current version,.

**Review by lawnjelly — APPROVED** [review]

**Comment by jordo** [comment]
> We discussed this PR a bit today in rocket chat:
> https://chat.godotengine.org/channel/devel/thread/ECRSuitLBW9BxSiTJ
> 
> and the general consensus was at least me and reduz were happy to try this out in the 3.4 betas. That is assuming it has been tested and appears to work as well as can be tested prior to a wider rollout. If there are any problems I'm guessing they will appear pretty soon.
> 
> In the long run we might do some improvements to the timing code / refactoring but this should be good to fix the bugs in the current version,.

thx @lawnjelly !  Sounds like a good appoach

**Comment by lawnjelly** [comment]
@zmanuel sorry there's been a bit of a delay on this, I think if you are able to rebase to fix that small double / float change this should be fine to merge and try out.

**Comment by zmanuel** [comment]
> @zmanuel sorry there's been a bit of a delay on this, I think if you are able to rebase to fix that small double / float change this should be fine to merge and try out.
Sure, done. And no worries.

Also, feel free to completely remove the physics_jitter_fix-code once your frame time smoothing solution is complete (maybe it already is? I admit I haven't been paying close attention, I just saw it in the release notes). This here has no significant practical advantages; it converges quicker to stable updates, that's about it. IMHO, that does not justify the really hard to follow logic with the typical_physics_steps.

**Comment by Calinou** [comment]
> once your frame time smoothing solution is complete (maybe it already is? I admit I haven't been paying close attention, I just saw it in the release notes).

It's not available yet; [it's planned for a future 4.x release](https://github.com/godotengine/godot-proposals/issues/2753).

**Comment by lawnjelly** [comment]
> > once your frame time smoothing solution is complete (maybe it already is? I admit I haven't been paying close attention, I just saw it in the release notes).
> 
> It's not available yet; [it's planned for a future 4.x release](https://github.com/godotengine/godot-proposals/issues/2753).

I think this is referring to #48390 rather than the timestep interpolation, me and @zmanuel discussed it on rocketchat a while ago.

While they may both solve the stairstep effect in some circumstances, I'm happy to get this PR in, and we can look at a later date at whether there is an argument to remove jitter fix.

At the moment the delta smoothing operates only in limited circumstances (vsynced, and when meeting vsync intervals), and jitter fix may still be helping in wider circumstances. We probably need to test this in more detail before deciding whether there is a route to moving away from jitter fix.

**Comment by lawnjelly** [comment]
The PR isn't showing as mergeable as it still thinks there are requested changes. @Calinou is there still something that needs to be done or is this a github hiccup? :thinking:

**Review by Calinou — APPROVED** [review]

**Comment by zmanuel** [comment]
> The PR isn't showing as mergeable as it still thinks there are requested changes. @Calinou is there still something that needs to be done or is this a github hiccup? thinking

There were some conversation items relating to outdated code changes that were considered unresolved; apparently I was able to mark them as resolved just now. Maybe those were the hangup.


> At the moment the delta smoothing operates only in limited circumstances (vsynced, and when meeting vsync intervals), and jitter fix may still be helping in wider circumstances.

I doubt jitter fix does anything useful for non-vsync, the traditional one with tearing. But yeah, it may stabilize timesteps with variable sync (G-Sync/FreeSync) if the framerate is semi-stable. But in either of those cases, physics jitter would not have been a problem anyway, and I see no reason one could not teach the delta smoothing to work in a limited fashion in those situations, too.

**Comment by lawnjelly** [comment]
It looks like everyone interested has had a chance to look this over (Akien, reduz etc) and it looks good to me for testing out in master imo, worst case if we get problems it can be reverted.

Thanks!

**Comment by akien-mga** [comment]
This would need a dedicated backport PR for `3.x` as the code has changed significantly (mostly parameter renames, but I prefer not to mess it up by eyeballing the changes while cherry-picking several PRs).

**Comment by zmanuel** [comment]
> This would need a dedicated backport PR for `3.x` as the code has changed significantly (mostly parameter renames, but I prefer not to mess it up by eyeballing the changes while cherry-picking several PRs).

Sure thing, here you go: #52947
I had an old version of the fix that still used the 3.x variable names around, that way I had something to cross-check.
